File: Sources/KakaoSapiens/Services/SyncAccountTransitionModel.swift
import Foundation

public enum SyncTransitionAvailability {
    case ready
    case syncEnabled
    case outboxPending
    case noActiveAccount
    case recoveryRequired
}

public protocol SyncAccountTransitionServicing {
    func availability() -> SyncTransitionAvailability
    func unlink() throws
}

public enum SyncAccountTransitionUIState: Equatable {
    case idle
    case ready
    case blocked
    case confirmingJoin
    case joining
    case confirmingUnlink
    case unlinking
    case unlinked
    case error
}

public struct SyncAccountTransitionActions: Equatable {
    public var canRequestJoin = false
    public var canConfirmJoin = false
    public var canStartJoin = false
    public var canRequestUnlink = false
    public var canConfirmUnlink = false
    public var canDismiss = false
}

@MainActor
public final class SyncAccountTransitionModel: ObservableObject {
    @Published public private(set) var state: SyncAccountTransitionUIState = .idle
    private let service: SyncAccountTransitionServicing

    public init(service: SyncAccountTransitionServicing) { self.service = service }

    public var actions: SyncAccountTransitionActions {
        switch state {
        case .ready:
            return SyncAccountTransitionActions(canRequestJoin: true, canRequestUnlink: true)
        case .confirmingJoin:
            return SyncAccountTransitionActions(canConfirmJoin: true, canDismiss: true)
        case .joining:
            return SyncAccountTransitionActions(canStartJoin: true, canDismiss: true)
        case .confirmingUnlink:
            return SyncAccountTransitionActions(canConfirmUnlink: true, canDismiss: true)
        case .error:
            return SyncAccountTransitionActions(canDismiss: true)
        default:
            return SyncAccountTransitionActions()
        }
    }

    public func refresh() {
        state = service.availability() == .ready ? .ready : .blocked
    }

    public func requestJoin() {
        if actions.canRequestJoin { state = .confirmingJoin }
    }

    public func confirmJoin() {
        if actions.canConfirmJoin { state = .joining }
    }

    public func requestUnlink() {
        if actions.canRequestUnlink { state = .confirmingUnlink }
    }

    public func confirmUnlink() {
        guard actions.canConfirmUnlink else { return }
        state = .unlinking
        do { try service.unlink(); state = .unlinked }
        catch { state = .error }
    }

    public func dismiss() {
        if actions.canDismiss { refresh() }
    }
}

File: android/app/src/main/java/com/sapiens/gagaodok/sync/SyncAccountTransitionModel.kt
package com.sapiens.gagaodok.sync

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

enum class SyncTransitionAvailability { READY, SYNC_ENABLED, OUTBOX_PENDING, NO_ACTIVE_ACCOUNT, RECOVERY_REQUIRED }

interface SyncAccountTransitionServicing {
    fun availability(): SyncTransitionAvailability
    fun unlink()
}

enum class SyncAccountTransitionUiState {
    IDLE, READY, BLOCKED, CONFIRMING_JOIN, JOINING, CONFIRMING_UNLINK,
    UNLINKING, UNLINKED, ERROR,
}

data class SyncAccountTransitionActions(
    val canRequestJoin: Boolean = false,
    val canConfirmJoin: Boolean = false,
    val canStartScanner: Boolean = false,
    val canRequestUnlink: Boolean = false,
    val canConfirmUnlink: Boolean = false,
    val canDismiss: Boolean = false,
)

class SyncAccountTransitionModel(private val service: SyncAccountTransitionServicing) {
    private val mutableState = MutableStateFlow(SyncAccountTransitionUiState.IDLE)
    val state: StateFlow<SyncAccountTransitionUiState> = mutableState.asStateFlow()

    val actions: SyncAccountTransitionActions
        get() = when (mutableState.value) {
            SyncAccountTransitionUiState.READY -> SyncAccountTransitionActions(
                canRequestJoin = true, canRequestUnlink = true,
            )
            SyncAccountTransitionUiState.CONFIRMING_JOIN -> SyncAccountTransitionActions(
                canConfirmJoin = true, canDismiss = true,
            )
            SyncAccountTransitionUiState.JOINING -> SyncAccountTransitionActions(
                canStartScanner = true, canDismiss = true,
            )
            SyncAccountTransitionUiState.CONFIRMING_UNLINK -> SyncAccountTransitionActions(
                canConfirmUnlink = true, canDismiss = true,
            )
            SyncAccountTransitionUiState.ERROR -> SyncAccountTransitionActions(canDismiss = true)
            else -> SyncAccountTransitionActions()
        }

    fun refresh() {
        mutableState.value = when (service.availability()) {
            SyncTransitionAvailability.READY -> SyncAccountTransitionUiState.READY
            else -> SyncAccountTransitionUiState.BLOCKED
        }
    }

    fun requestJoin() {
        if (actions.canRequestJoin) mutableState.value = SyncAccountTransitionUiState.CONFIRMING_JOIN
    }

    fun confirmJoin() {
        if (actions.canConfirmJoin) mutableState.value = SyncAccountTransitionUiState.JOINING
    }

    fun requestUnlink() {
        if (actions.canRequestUnlink) mutableState.value = SyncAccountTransitionUiState.CONFIRMING_UNLINK
    }

    fun confirmUnlink() {
        if (!actions.canConfirmUnlink) return
        mutableState.value = SyncAccountTransitionUiState.UNLINKING
        mutableState.value = try {
            service.unlink()
            SyncAccountTransitionUiState.UNLINKED
        } catch (_: Throwable) {
            SyncAccountTransitionUiState.ERROR
        }
    }

    fun dismiss() {
        if (actions.canDismiss) refresh()
    }
}
